### DISCUSSION TITLE
Avoiding failures in generating mappers for nested objects without any properties

### DISCUSSION THREAD

**Discussion by sachingoel0101** [mapstruct/mapstruct#3627] [Q&A]
I have a deeply nested structure of objects, not all of which I can control. I am trying to use Mapper at the very top level object, and let MapStruct automatically generate nested mappings. However, I run into issues with nested objects here and there, which do not contain any properties (and consequently, getters/setters). 

As an example,

```
@Data
public class A {

    String x;
    ABC abc;
}

class ABC {}
```

```
@Data
public class AInternal {

    String x;
    ABCInternal abc;
}

class ABCInternal {}
```

```
@org.mapstruct.Mapper()
public interface Mapper {

    AInternal convert(A a);

    A converter(AInternal aInternal);

}
```

fails with ```No target bean properties found: can't map property "ABCInternal abc" to "ABC abc". Consider to declare/implement a mapping method: "ABC map(ABCInternal value)".```

I have considered declaring custom mapper methods for these empty objects, however - 1) There are almost 30-40 such objects, and I cannot write a custom dummy mapper for each, 2) New fields/properties may get introduced to these objects at any point, and I'll have to constantly keep removing the dummy mappers. 

Is there a MapStruct configuration that allows such empty objects to be handled as expected? 

[I would like to use unmappedSourcePolicy/unmappedTargetPolicy as Error to ensure that the properties that do exist must be mapped properly].

**Comment by zyberzebra** [comment]
>   New fields/properties may get introduced to these objects at any point,

The Problem/Failure you get only happens for those “empty” objects, if I'm not mistaken. What I don't understand is, why are there empty objects at all? What is their purpose? 

---

You could “collect” all your empty Objectmappings in a different mapper and us that one in your “main”-Mapper:
```java
@org.mapstruct.Mapper(
        uses = EmptyObjectMapper.class
)
public interface Mapper {
    AInternal convert(A a);
    A convert(AInternal aInternal);
}

@org.mapstruct.Mapper
interface EmptyObjectMapper {
    ABCInternal convert (ABC aInternal);
    ABC convert (ABCInternal aInternal);
}
``` 

This would lead to :

If ABCInternal != null a new ABC will be initialized, if ABCInternal == null, ABC will be null and vice versa.

If you want your ABCInternal or/and ABC to be always null or initialized, you could always do default implementations:

```java
@org.mapstruct.Mapper
interface EmptyObjectMapperDefault {
    default ABCInternal convert (ABC aInternal){
        return null;     // return new ABCInternal();
    }
    default ABC convert (ABCInternal aInternal){
        return null;     // return new ABC();
    }
}
```
